### PR TITLE
Update QROM decomposition to use more optimal resources

### PR DESCRIPTION
**Pull request by ddhawan11**
**Context:**
QROM and Adjoint QROM use a better decomposition based on measurement based techniques.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-104106]
[sc-105143]

### CODE REVIEW

File: pennylane/estimator/templates/subroutines.py [hunk 1]
```diff
@@ -1641,6 +1641,9 @@ class QROM(ResourceOperator):
         :code:`restored = False` and `Berry et al. (2019) <https://arxiv.org/pdf/1902.02134>`_
         (Figure 4) for :code:`restored = True`.
 
+        For :code:`restored = False`, we further use the measurement based technique described in
+        `arXiv:1092.02134 <https://arxiv.org/abs/1902.02134>_` to optimize the cost.
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
        For :code:`restored = False`, we use the measurement based technique described in
        `arXiv:1092.02134 <https://arxiv.org/abs/1902.02134>_` to further optimize the cost.
```

Comment (austingmhuang) [reply pennylane/estimator/templates/subroutines.py]: And so on and so forth.

Comment (austingmhuang) [reply pennylane/estimator/templates/subroutines.py]: Can also just delete "further" altogether.

File: pennylane/estimator/templates/subroutines.py [hunk 2]
```diff
@@ -1823,6 +1829,9 @@ def resource_decomp(
 
             if restored:
                 gate_cost.append(Deallocate((W_opt - 1) * size_bitstring))  # release Swap registers
+            else:
+                gate_cost.append(GateCount(x, (W_opt - 1) * size_bitstring))  # measure and reset
+                gate_cost.append(Deallocate((W_opt - 1) * size_bitstring))  # release swap registers
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]: Seems like either way there's a Deallocate?

```suggestion
            if not restored:
                gate_cost.append(GateCount(x, (W_opt - 1) * size_bitstring))  # measure and reset
            gate_cost.append(Deallocate((W_opt - 1) * size_bitstring))  # release Swap registers
```

something like this?

File: pennylane/estimator/templates/subroutines.py [hunk 3]
```diff
@@ -1998,6 +2007,146 @@ def resource_params(self) -> dict:
             "restored": self.restored,
         }
 
+    @classmethod
+    def _ctrl_T(cls, L_opt, num_bit_flips, count=1):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]: Please provide type hints for this function

File: pennylane/estimator/templates/subroutines.py [hunk 4]
```diff
@@ -1998,6 +2007,146 @@ def resource_params(self) -> dict:
...
+        else:
+            gate_cost.append(
+                GateCount(
+                    x, count * num_bit_flips
+                )  # each unitary in the select is just an X gate to load the data
+            )
+        return gate_cost
+
+    @classmethod
+    def _ctrl_S(cls, num_ctrl_wires, count=1):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]: Please provide type hints for this function

File: pennylane/estimator/templates/subroutines.py [hunk 5]
```diff
@@ -1998,6 +2007,146 @@ def resource_params(self) -> dict:
...
+        """Constructs the control-S subroutine as defined in Figure 8 of
+        `arXiv:1092.02134 <https://arxiv.org/abs/1902.02134>_` excluding the initial X gate."""
+        num_ctrl_swaps = 2**num_ctrl_wires - 1
+        return [qre.GateCount(qre.resource_rep(qre.CSWAP), count * num_ctrl_swaps)]
+
+    @classmethod
+    def _ctrl_S_adj(cls, num_ctrl_wires, count=1):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]: Please provide type hints for this function

File: pennylane/estimator/templates/subroutines.py (hunk 3 above)
Comment (austingmhuang) [reply pennylane/estimator/templates/subroutines.py]: Unclear what L_opt is.

File: tests/estimator/templates/test_estimator_subroutines.py [hunk 6]
```diff
@@ -1663,6 +1667,173 @@ def test_controlled_res_decomp(
...
+    @pytest.mark.parametrize(
+        "num_data_points, output_size, depth",
+        (
+            (100, 10, 2),
+            (100, 2, 4),
+            (12, 1, 1),
+        ),
+    )
+    def test_toffoli_counts(self, num_data_points, output_size, depth):
+        """Test that the Toffoli counts are correct compared to the paper."""
```
Comment (austingmhuang) [inline tests/estimator/templates/test_estimator_subroutines.py]: I would link the paper here as a comment.

File: tests/estimator/templates/test_estimator_subroutines.py [hunk 7]
```diff
@@ -1423,12 +1425,14 @@ def test_resource_rep(self, num_data_points, size_data_points, num_bit_flips, de
                 12,
                 2,
                 5,
-                128,  # This will get turncated to 16 as the max depth
+                128,  # This will get truncated to 16 as the max depth
```
Comment (austingmhuang) [inline tests/estimator/templates/test_estimator_subroutines.py:1548]: 👍

**Review by austingmhuang — COMMENTED** [review]
Some initial comments.

File: pennylane/estimator/templates/subroutines.py [hunk 8]
```diff
@@ -1998,6 +2005,172 @@ def resource_params(self) -> dict:
             "restored": self.restored,
         }
 
+    @classmethod
+    def _ctrl_T(cls, num_data_blocks: int, num_bit_flips: int, count=1):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
    def _ctrl_T(cls, num_data_blocks: int, num_bit_flips: int, count=1) -> List[GateCount]:
```

File: pennylane/estimator/templates/subroutines.py [hunk 9]
```diff
@@ -1998,6 +2005,172 @@ def resource_params(self) -> dict:
...
+        else:
+            gate_cost.append(
+                GateCount(
+                    x, count * num_bit_flips
+                )  # each unitary in the select is just an X gate to load the data
+            )
+        return gate_cost
+
+    @classmethod
+    def _ctrl_S(cls, num_ctrl_wires: int, count: int = 1):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
    def _ctrl_S(cls, num_ctrl_wires: int, count: int = 1) -> List[GateCount]:
```

File: pennylane/estimator/templates/subroutines.py [hunk 10]
```diff
@@ -1998,6 +2005,172 @@ def resource_params(self) -> dict:
...
+            count (int): The number of times to repeat the subroutine.
+
+        Returns:
+            list[:class:`~.pennylane.estimator.resource_operator.GateCount`]: The resource decomposition of the control- :math:`S` subroutine.
+        """
+        num_ctrl_swaps = 2**num_ctrl_wires - 1
+        return [qre.GateCount(qre.resource_rep(qre.CSWAP), count * num_ctrl_swaps)]
+
+    @classmethod
+    def _ctrl_S_adj(cls, num_ctrl_wires: int, count: int = 1):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
    def _ctrl_S_adj(cls, num_ctrl_wires: int, count: int = 1) -> List[GateCount]:
```

File: pennylane/estimator/templates/subroutines.py [hunk 11]
```diff
@@ -1998,6 +2005,172 @@ def resource_params(self) -> dict:
...
+
+        num_ops = 2**num_ctrl_wires - 1
+        return [
+            qre.GateCount(h, count * num_ops),
+            qre.GateCount(cz, count * num_ops),
+            qre.GateCount(cnot, count * num_ops),
+        ]
+
+    @classmethod
+    def adjoint_resource_decomp(cls, target_resource_params: dict):
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
    def adjoint_resource_decomp(cls, target_resource_params: dict) -> List[GateCount]:
```

File: pennylane/estimator/templates/subroutines.py [hunk 12]
```diff
@@ -1753,6 +1756,9 @@ def resource_decomp(
             :code:`restored = False` and `Berry et al. (2019) <https://arxiv.org/pdf/1902.02134>`_
             (Figure 4) for :code:`restored = True`.
 
+            For :code:`restored = False`, we use the measurement based technique described in
+            `arXiv:1092.02134 <https://arxiv.org/abs/1902.02134>_` to further optimize the cost.
+
```
Comment (Jaybsoni) [inline pennylane/estimator/templates/subroutines.py:2137]: I think these can be merged together. We are already using this paper for the `restored == True` so we should just link the one paper.

File: pennylane/estimator/templates/subroutines.py [hunk 13]
```diff
@@ -1998,6 +2005,172 @@ def resource_params(self) -> dict:
...
+        num_ops = 2**num_ctrl_wires - 1
+        return [
+            qre.GateCount(h, count * num_ops),
+            qre.GateCount(cz, count * num_ops),
+            qre.GateCount(cnot, count * num_ops),
+        ]
+
+    @classmethod
+    def adjoint_resource_decomp(cls, target_resource_params: dict) -> list[GateCount]:
```
Comment (Jaybsoni) [inline pennylane/estimator/templates/subroutines.py:2471]: Can you double check that the output makes sense in the edge case where select_swap_depth = 1 for both restored = True and restored = False. Can you add a test case for this.

File: pennylane/estimator/templates/subroutines.py (hunk 12 above)
Comment (ddhawan11) [reply pennylane/estimator/templates/subroutines.py:2137]: I think we should still link the Low et al paper, modified it to clean things up and combine where needed.

**Review by austingmhuang — COMMENTED** [review]
Although it's not really in the scope of the PR, the typehints in the docstring for the QROM and QROM's resource methods are inaccurate.

Should be something like this:
```
Parameters:
num_bitstrings (int) – the number of bitstrings that are to be encoded

size_bitstring (int) – the length of each bitstring

num_bit_flips (int | None) – The total number of 1’s in the dataset. Defaults to (num_bitstrings * size_bitstring) // 2, which is half the dataset.

restored (bool) – Determine if allocated qubits should be reset after the computation (at the cost of higher gate counts). Defaults to True.

select_swap_depth (int | None) – A parameter 𝜆 that determines if data will be loaded in parallel by adding more rows following Figure 1.C of [Low et al. (2024)](https://arxiv.org/pdf/1812.00954). Can be None, 1 or a positive integer power of two. Defaults to None, which internally determines the optimal depth.

wires (WiresLike, None) – The wires the operation acts on (control and target). Excluding any additional qubits allocated during the decomposition (e.g select-swap wires).
```

We should also check that the docs are building properly. For me I see two things that don't look right.

Screenshot 2025-12-19 at 9 53 20 AM

Screenshot 2025-12-19 at 9 52 48 AM

Please double check.

File: pennylane/estimator/templates/subroutines.py [hunk 14]
```diff
@@ -1726,32 +1731,37 @@ def __init__(
...
+            num_bit_flips (int | None): The total number of :math:`1`'s in the dataset. Defaults to
                 :code:`(num_bitstrings * size_bitstring) // 2`, which is half the dataset.
             select_swap_depth (int | None): A parameter :math:`\lambda` that determines
                 if data will be loaded in parallel by adding more rows following Figure 1.C of
                 `Low et al. (2024) <https://arxiv.org/pdf/1812.00954>`_. Can be :code:`None`,
                 :code:`1` or a positive integer power of two. Defaults to :code:`None`, which internally
                 determines the optimal depth.
-            restored (bool, optional): Determine if allocated qubits should be reset after the computation
+            restored (bool): Determine if allocated qubits should be reset after the computation
                 (at the cost of higher gate counts). Defaults to :code`True`.
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
            restored (bool): Determine if allocated qubits should be reset after the computation
                (at the cost of higher gate counts). Defaults to :code:`True`.
```

File: pennylane/estimator/templates/subroutines.py [hunk 15]
```diff
@@ -1998,24 +2009,190 @@ def resource_params(self) -> dict:
...
+    @classmethod
+    def adjoint_resource_decomp(cls, target_resource_params: dict) -> list[GateCount]:
+        r"""Returns a list representing the resources of the adjoint of the operator. Each object represents a quantum gate
+        and the number of times it occurs in the decomposition.
+
+        Args:
+            target_resource_params(dict): A dictionary containing the resource parameters of the target operator.
+
+        Resources:
+            This resources are based on Appendix C of `arXiv:1092.02134 <https://arxiv.org/abs/1902.02134>_`.
```
Comment (austingmhuang) [inline pennylane/estimator/templates/subroutines.py]:
```suggestion
            This resources are based on Appendix C of `arXiv:1092.02134<https://arxiv.org/abs/1902.02134>_`.
```

**Review by Jaybsoni — APPROVED** [review]
Looks good to go 👍🏼
